### PR TITLE
[FLINK-12660][hive] Integrate Flink with Hive UDAF

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

This PR adds support for Hive UDAF in Flink.

For Hive UDAF, there are mainly four interfaces: UDAF, GenericUDAFResolver, GenericUDAFResolver2, AbstractGenericUDAFResolver

UDAF is kind of on its own, deprecated in Hive a long time ago. GenericUDAFResolver and AbstractGenericUDAFResolver are deprecated for GenericUDAFResolver2.

## Brief change log

- added HiveGenericUDAF to wrap Hive UDAF and GenericUDAFResolver2
- added unit tests

## Verifying this change

This change added tests and can be verified as follows: `HiveGenericUDAFTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by bowenli86** [comment]
cc @xuefuz @lirui-apache @zjuwangg

**Comment by bowenli86** [comment]
@flinkbot attention @xuefuz @JingsongLi

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDAF.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+	private Object[] constantArguments;
+	private DataType[] argTypes;
+
+	private transient GenericUDAFEvaluator partialEvaluator;
+	private transient GenericUDAFEvaluator finalEvaluator;
+	private transient ObjectInspector finalResult;
+	private transient HiveObjectConversion[] conversions;
+	private transient boolean allIdentityConverter;
+	private transient boolean inited;
```
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDAF.java]: Nit: let's spelled out the word.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDAF.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		for (int i = 0; i < inputInspectors.length; i++) {
+			conversions[i] = HiveInspectors.getConversion(inputInspectors[i], argTypes[i].getLogicalType());
+		}
+		allIdentityConverter = Arrays.stream(conversions)
+			.allMatch(conv -> conv instanceof IdentityConversion);
+
+		inited = true;
+	}
+
+	private GenericUDAFEvaluator newEvaluator(ObjectInspector[] inputInspectors) throws SemanticException {
```
Comment (xuefuz) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDAF.java]: Nit: consider renaming the method.

**Comment by bowenli86** [comment]
@xuefuz thanks for the feedback! I've addressed them

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDAF.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		if (isUDAFBridgeRequired) {
+			resolver = new GenericUDAFBridge((UDAF) hiveFunctionWrapper.createFunction());
+		} else {
+			resolver = (GenericUDAFResolver2) hiveFunctionWrapper.createFunction();
+		}
+
+		return resolver.getEvaluator(
+			new SimpleGenericUDAFParameterInfo(
+				inputInspectors,
+				// The flag to indicate if the UDAF invocation was from the windowing function call or not.
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveGenericUDAF.java]: 👍 Detailed comments.
Hive UDAF

**Comment by bowenli86** [comment]
@xuefuz @JingsongLi thanks for your review!

The UT failure in Travis is unrelated.

Merging

**Pull request closed without merging**